restore default port 8500 for CONSUL_HTTP_ADDR (#394)

larrytamnjong opened the pull request:
Restore default port 8500 for CONSUL_HTTP_ADDR when no port is specif…ied (#394)

larrytamnjong commented:
@marcin-krystianc I made a tiny adjustment to default the port to 8500 if it isn’t specified in CONSUL_HTTP_ADDR. Please let me know if this makes sense.

marcin-krystianc reviewed (commented):
Hi @larrytamnjong,

implementation seems legit, please add tests for following cases:
```
CONSUL_HTTP_ADDR=127.0.0.1 => protocol: http, host: 127.0.0.1, port: 8500
CONSUL_HTTP_ADDR=http://127.0.0.1/ => protocol: http, host: 127.0.0.1, port: 8500 
CONSUL_HTTP_ADDR=http://127.0.0.1:8500/ => protocol: http, host: 127.0.0.1, port: 8500
CONSUL_HTTP_ADDR=http://127.0.0.1:80/ => protocol: http, host: 127.0.0.1, port: 80
```

Please also verify the test results are the same for your implementation and the version before the https://github.com/G-Research/consuldotnet/commit/ef310c22d35684892bd2231a53f3e9bbf907c78a.

larrytamnjong commented:
Hi @marcin-krystianc 
I create a test method and for my implementation 
```
 public void ConsulClient_BaseUri_Parses_ConsulHttpAddr_Correctly()
 {
     Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "127.0.0.1");
     var client1 = new ConsulClient();
     var config1 = client1.Config;


     Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1/");
     var client2 = new ConsulClient();
     var config2 = client2.Config;


     Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1:8500/");
     var client3 = new ConsulClient();
     var config3 = client3.Config;


     Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1:80/");
     var client4 = new ConsulClient();
     var config4 = client4.Config;

     Assert.Equal("127.0.0.1", config1.Address.Host);  //Passes
     Assert.Equal("127.0.0.1", config2.Address.Host);  //Passes
     Assert.Equal("127.0.0.1", config4.Address.Host);  //Passes
     Assert.Equal("127.0.0.1", config3.Address.Host); //Passes
     Assert.Equal(8500, config1.Address.Port);  //Passes
     Assert.Equal(8500, config2.Address.Port); //Passes
     Assert.Equal(8500, config3.Address.Port);  //Passes
     Assert.Equal(80, config4.Address.Port);  //It fails here because the port returns 8500
 }
```
However, when I run this test on commit f6a8038, I encounter exceptions with certain values of CONSUL_HTTP_ADDR:
```
 public void ConsulClient_BaseUri_Parses_ConsulHttpAddr_Correctly()
 {
     Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "127.0.0.1");
     var client1 = new ConsulClient();
     var config1 = client1.Config;


     Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1/");  //An exception is thrown here "Failed parsing port from environment variable CONSUL_HTTP_ADDR"
     var client2 = new ConsulClient();
     var config2 = client2.Config;


     Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1:8500/"); //An exception is thrown here "Failed parsing port from environment variable CONSUL_HTTP_ADDR"
     var client3 = new ConsulClient();
     var config3 = client3.Config;


     Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1:80/"); //An exception is thrown here "Failed parsing port from environment variable CONSUL_HTTP_ADDR"
     var client4 = new ConsulClient();
     var config4 = client4.Config;

     Assert.Equal("127.0.0.1", config1.Address.Host);  
     Assert.Equal("127.0.0.1", config2.Address.Host);  
     Assert.Equal("127.0.0.1", config4.Address.Host);  
     Assert.Equal("127.0.0.1", config3.Address.Host); 
     Assert.Equal(8500, config1.Address.Port);  
     Assert.Equal(8500, config2.Address.Port); 
     Assert.Equal(8500, config3.Address.Port);  
     Assert.Equal(80, config4.Address.Port);  
 }
```
I'm not sure if this is the expected way to test it. Could you provide some guidance on how to proceed?

orange-puff commented:
This seems to not be quite right. `Uri.IsDefaultPort` will return true if the specified port matches the scheme. So, 
`new Uri("http://mysite.com").IsDefaultPort` and `new Uri("http://mysite.com:80").IsDefaultPort` will both resolve to `true`.
It seems dumb, but what if the consumer has their consul agent listening on port `80`? This logic would overwrite their port with `8500`

For determining if the original string contains a port, I'd suggest either

`envAddr.Contains($"{uri.Host}:{uri.Port}");`

or
`Regex.IsMatch(envAddr, ":[0-9]+");`

orange-puff commented:
It looks like https://github.com/G-Research/consuldotnet/commit/f6a8038af144ffa3a8ebf8b05cb3181aad4c50f6 breaks if the client specifies a scheme `http` or `https`, or any other for that matter. There is an environment variable driven way of determining scheme

larrytamnjong commented:
> This seems to not be quite right. `Uri.IsDefaultPort` will return true if the specified port matches the scheme. So, `new Uri("http://mysite.com").IsDefaultPort` and `new Uri("http://mysite.com:80").IsDefaultPort` will both resolve to `true`. It seems dumb, but what if the consumer has their consul agent listening on port `80`? This logic would overwrite their port with `8500`
> 
> For determining if the original string contains a port, I'd suggest either
> 
> `envAddr.Contains($"{uri.Host}:{uri.Port}");`
> 
> or `Regex.IsMatch(envAddr, ":[0-9]+");`

This makes sense

marcin-krystianc commented:
> Hi @marcin-krystianc I create a test method and for my implementation
> 
> ```
>  public void ConsulClient_BaseUri_Parses_ConsulHttpAddr_Correctly()
>  {
>      Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "127.0.0.1");
>      var client1 = new ConsulClient();
>      var config1 = client1.Config;
> 
> 
>      Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1/");
>      var client2 = new ConsulClient();
>      var config2 = client2.Config;
> 
> 
>      Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1:8500/");
>      var client3 = new ConsulClient();
>      var config3 = client3.Config;
> 
> 
>      Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1:80/");
>      var client4 = new ConsulClient();
>      var config4 = client4.Config;
> 
>      Assert.Equal("127.0.0.1", config1.Address.Host);  //Passes
>      Assert.Equal("127.0.0.1", config2.Address.Host);  //Passes
>      Assert.Equal("127.0.0.1", config4.Address.Host);  //Passes
>      Assert.Equal("127.0.0.1", config3.Address.Host); //Passes
>      Assert.Equal(8500, config1.Address.Port);  //Passes
>      Assert.Equal(8500, config2.Address.Port); //Passes
>      Assert.Equal(8500, config3.Address.Port);  //Passes
>      Assert.Equal(80, config4.Address.Port);  //It fails here because the port returns 8500
>  }
> ```
> 
> However, when I run this test on commit [f6a8038](https://github.com/G-Research/consuldotnet/commit/f6a8038af144ffa3a8ebf8b05cb3181aad4c50f6), I encounter exceptions with certain values of CONSUL_HTTP_ADDR:
> 
> ```
>  public void ConsulClient_BaseUri_Parses_ConsulHttpAddr_Correctly()
>  {
>      Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "127.0.0.1");
>      var client1 = new ConsulClient();
>      var config1 = client1.Config;
> 
> 
>      Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1/");  //An exception is thrown here "Failed parsing port from environment variable CONSUL_HTTP_ADDR"
>      var client2 = new ConsulClient();
>      var config2 = client2.Config;
> 
> 
>      Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1:8500/"); //An exception is thrown here "Failed parsing port from environment variable CONSUL_HTTP_ADDR"
>      var client3 = new ConsulClient();
>      var config3 = client3.Config;
> 
> 
>      Environment.SetEnvironmentVariable("CONSUL_HTTP_ADDR", "http://127.0.0.1:80/"); //An exception is thrown here "Failed parsing port from environment variable CONSUL_HTTP_ADDR"
>      var client4 = new ConsulClient();
>      var config4 = client4.Config;
> 
>      Assert.Equal("127.0.0.1", config1.Address.Host);  
>      Assert.Equal("127.0.0.1", config2.Address.Host);  
>      Assert.Equal("127.0.0.1", config4.Address.Host);  
>      Assert.Equal("127.0.0.1", config3.Address.Host); 
>      Assert.Equal(8500, config1.Address.Port);  
>      Assert.Equal(8500, config2.Address.Port); 
>      Assert.Equal(8500, config3.Address.Port);  
>      Assert.Equal(80, config4.Address.Port);  
>  }
> ```
> 
> I'm not sure if this is the expected way to test it. Could you provide some guidance on how to proceed?

@larrytamnjong , we have already a parametrised test `Client_DefaultConfig_env`, just add more cases to that test instead of implementing a new test method. 
As per @orange-puff comment, the old version didn't support parsing of all possible combination of scheme, address and port so in such cases we do not have to worry about backwards compatibility. What we  care about is that if an `CONSUL_HTTP_ADDR` value was valid in the old version it should be also valid in the new version and it should produce the same consul URL.

larrytamnjong commented:
Hi @marcin-krystianc,

I've added tests for the scenarios you mentioned and adjusted the check for the default port. The tests pass in both my implementation and commit f6a8038. The main difference is that in previous versions, we can't specify an HTTP scheme when adding the address, as noted by @orange-puff.

I believe it’s ready now. Let me know your thoughts!

marcin-krystianc commented on Consul/Client.cs:
```diff
@@ -174,8 +174,14 @@ private void ConfigureFromEnvironment(UriBuilder consulAddress)
+                {
+                    consulAddress.Port = uri.Port;
+                }
+                else
+                {
+                    consulAddress.Port = 8500;
+                }
```
Assigning the default 8500 here is redundant as the `consulAddress` variable already has the default port pre-assigned. You can remove the `else` block entirely.

marcin-krystianc commented on Consul/Client.cs:
```diff
@@ -174,8 +174,14 @@ private void ConfigureFromEnvironment(UriBuilder consulAddress)
                 }
-
-                consulAddress.Port = uri.Port;
+                if (envAddr.Contains($"{uri.Host}:{uri.Port}"))
```
👍 I'm not a fan of this, but it does the job and  I don't see how we can do it differently.

larrytamnjong replied:
:)
